### ISSUE TITLE
Create [TrustedTypes].fromLiteral method

### ISSUE THREAD

**Issue by shhnjk** [w3c/webappsec-trusted-types#347]
Requirement of Trusted Types policy for assigning static HTML, Script, and ScriptURL is a pain point of TT conversion.
Creating some element [using DOM APIs](https://source.chromium.org/chromium/chromium/src/+/main:out/android-Debug/gen/ui/webui/resources/preprocessed/js/cr/ui/tree.js;l=303-317;drc=4964b3de494f45977362824aa4f9167d0cf283dc) are hard, and creating policy just for [assigning a static ScriptURL](https://source.chromium.org/chromium/chromium/src/+/main:components/about_ui/resources/about_credits.js;l=15-17;drc=7ebdf7c9b0f8b5496ea4d4c4a4838965ec04a121) increases work and decreases readability without (almost) any benefit.

In order to make migration easier for website to increase adaptation of Trusted Types, safe code should require much less work to migrate.

While [Array.isTemplateObject](https://github.com/tc39/proposal-array-is-template-object/) aims to provide this primitive, it's unsure if that's landing anywhere, and that won't eliminate the requirement of TT Policy.

Instead, can we create a `fromLiteral` method under each Trusted Types, so that staticness can be verified in the Web Platform (using same primitive as `Array.isTemplateObject`)?

Example would be:
```
const s = TrustedHTML.fromLiteral`<s>Trusted</s>`; // returns TrustedHTML
node.innerHTML = s; // No TT violation as TrustedHTML was assigned
```

**Comment by Sora2455** [comment]
In my uninformed opinion, this could easily be a footgun used by server-side templating languages to bypass the protection offered by Trusted Types. I could easily see something like this happening:

```
loginDescripter.innerHTML = TrustedHTML.fromLiteral`<i><?php echo $username;?></i>`;
```

**Comment by koto** [comment]
Trusted Types are explicitly [not covering server-side injections](https://w3c.github.io/webappsec-trusted-types/dist/spec/#non-goals). Without `fromLiteral`, what prevents the developers from just outputting arbitrary strings in the HTML response, or interpolating into the TT policy body?

**Comment by shhnjk** [comment]
@koto, what's your thought on this?

**Comment by domenic** [comment]
Strategically, I think this kind of thing is better than waiting on Array.isTemplateObject.

No existing web API defines a template tag function. Normally they would define it using the IDL `fromLiteral(sequence<DOMString> strings, any... args)`, but I guess then you would not be able to perform the verification you want. So indeed you should use `object`, perform the verification, and then use prose to convert to `sequence<DOMString>`.

Peeking into ES slots is discouraged in general---Web IDL is supposed to provide a uniform interface for interfacing with JS objects. But doing so makes sense in some cases, and IMO this is one of them. If it comes up more than a few times then we should define a wrapper algorithm in the IDL spec (e.g., the ["underlying buffer"](https://heycam.github.io/webidl/#buffersource-underlying-buffer) wrapper algorithm abstracts away the process of looking at the [[ViewedArrayBuffer]] internal slot.)

One thing I didn't see mentioned is how you plan to do the interpolation? E.g. if someone does

```js
const foo = `<script>console.log(1)</script>`
const s = TrustedHTML.fromLiteral(`<s>${foo}</s>`);
```

what is the result? Since you control the tag you could make the policy be that any interpolations involve HTML escaping, i.e. the result would be a trusted HTML string `<s>&lt;script>console.log(1)&lt/script></s>`. Is that the intention?

**Comment by koto** [comment]
Thanks @domenic, this seems workable :)

>  how you plan to do the interpolation?

At least here, the intention is to throw on any interpolation attempt (i.e. the template array will be of length 1). Template literals are just a way to enforce constant literals that we know are safe in a TT model (i.e. they are not user-injected). 

I would love to eventually allow for e.g. contextually escape HTML, but I'm not sure TT (or any web API to be honest) would be the right way to implement it. lit-html already does it, so we know it's feasible in a lib, and it's probably better suited there.

**Comment by domenic** [comment]
Interesting. I would have thought that just escaping all of `<`, `>`, `"`, `'`, and `&` would be good enough. (In the sense that it would definitely be "secure", and if it over-interpolates for some use cases, then oh well, people can just not use it.) But I am not an expert.

**Comment by koto** [comment]
Nope :) ```TrustedHTML`<${'script'}>'boom!'` ``` or ```TrustedHTML`<div ${'onclick'}="${'boom!'}">` ```. Strings are just bad without knowing the exact context.

**Comment by koto** [comment]
> > 1. It offers a way to create TT instances, so, like policies, it should be guardable by `trusted-types` directive (so - keyword). I can imagine how it's undesireable for developers to load scripts from arbitrary URLs, even in an absence of `script-src`.
> 
> If this requires policy, then I'm not sure why we don't require policy for [setHTML](https://github.com/WICG/sanitizer-api/issues/100).

To clarify, I meant guards, and not TT policy. For example, a keyword in `trusted-types` that can enable/disable `fromLiteral. I feel like we might be losing a lot control - TT allow to put guards on script loading mechanisms done by large codebases that don't share the problems of `script-src`. 

`setHTML`, as I understand it, currently sanitizes, and this is a non-sanitizing (but only accepting constants) equivalent? 
 
> > 2. I see a clear advantage - it's both secure, and easier to use than policies for these limited cases. I'm not sure how much in practice this buys, given that this can already be done (modulo foolproofness of `Array.isTemplateObject`) in a library that internally uses a policy (in fact, we already have two such libraries). The API could only do this securely if there's no interpolation whatsoever, whereas the library can be configurable (e.g. allow dynamic hostnames for a script URL, but only for these  hostnames, etc.). The libraries are always easier to experiment on the API shape of, and quicker to iterate on than the Web APIs. The libraries also could workaround transpilation to ES version that does not support template literals, whereas an API cannot.
> 
> _IF_ we can have this primitive without a policy, there are few advantages over library.
> 
> 1. We can use this mechanism in migration of third-party library (if keyword is required, this method will likely not used in third-party library).

I see. It could be an opt-out keyword though.

**Comment by shhnjk** [comment]
> I see. It could be an opt-out keyword though.

Yup, opt-out sounds great!

**Comment by shhnjk** [comment]
Assuming this will be an opt-out keyword, how should we go about designing what opt-out looks like?

My suggestion would be to create a built-in policy (just like `default`) where an application developer can create a policy specific to output of `fromLiteral`.

Example:
```
<meta http-equiv="Content-Security-Policy" content="require-trusted-types-for 'script'; trusted-types deny-literals;">

<script>
  trustedTypes.createPolicy('deny-literals', {
    // HTML and Script are fine
    createHTML: html => html,
    createScript: script => script,
    // Let's make sure script is same-origin
    createScriptURL: url => {
      const script_url = new URL(url);
      if (globalThis.origin === script_url.origin) {
        return script_url;
      }
      throw("Bad script URL");
    }
  });
</script>
```

**Comment by koto** [comment]
I'm confused - I thought the whole advantage of `fromLiteral` is that one doesn't need to create a policy at all :) 

A single, global policy - like a default policy - looks attractive, but is problematic - the model breaks if you integrate two codebases that use it in a single document. 

Policy in a `fromLiteral` introduces an additional concern. Tag functions are supposed not to have side effects, and policy functions can return different values every time, or alter global state. 

If rules for conversions are needed, I'd rather we keep using the regular policy API , as that one allows devs to have local rules (e.g. per module). We could, for example, change the first argument of `TrustedTypePolicy.prototype.createXYZ` functions to `any`, such that it can be called as a template literal:

```javascript
const foo = trustedTypes.createPolicy('literalsOnly', { 
  createScriptURL(template, ...args) {
     // in the future - Array.isTemplateObject
     if (!Array.isArray(template) || !Array.isArray(template.raw) || template.length !== 1) {
        throw 'bad';
    }
    return template[0];
  }
});

foo.createScriptURL`https://foo.bar`
```

`TrustedXYZ.fromLiteral` is, I think, only possible as a native API if there are no custom rules (due to side effects).

**Comment by shhnjk** [comment]
I do want `fromLiteral` without a policy. But I was thinking maybe we should give a context for folks who opt-out of `fromLiteral`, in such a way that they know something came to default policy (or some other policy) because they opted-out from `fromLiteral`. But given it can't have side effects, it makes sense that we can't do this.

In that case, we can simply let `fromLiteral` return a string when an app opt-out from `fromLiteral`?

**Comment by koto** [comment]
> can we simply let fromLiteral return a string when an app opt-out from fromLiteral?

I would rather avoid returning different return types from a function, API like that is hard to define types for, and requires either explicit type checking at the callsite, or leads to functional bugs.

For example, devs may expect getting a string and call `String.prototype.replace` on it, whereas code returns a `TrustedHTML` instead on some browsers, and this breaks. This is in fact exactly what happened with `DOMPurify.sanitize` and its TT support (which resulted in functional breakage). 

I'm afraid that, to avoid such errors, at most we should be returning `null` or `undefined` (but, actually, we should be throwing instead). The code to support it though is verbose, and requires checking for support separately :/

I changed my mind - if `fromLiteral` is safe (it is), it should "just" be always available, without any toggling via CSP. Folks who would wish to limit it, can just `delete TrustedHTML.fromLiteral`.

That way a "best effort" return value would be:

```javascript
foo.innerHTML = (TrustedHTML?.fromLiteral || (s => s[0]))`<p>this will be a TrustedHTML, or a string</p>`

// You can also store the tag function separately, and e.g. add some more checks for strings.

const bestLiteral = TrustedHTML?.fromLiteral || (s => {
  if (s.length > 1) {
    console.log('Please don't interpolate here'); // You can also throw.
  } 
  return s[0]
});

bestLiteral`<p>ha!</p>`
```

**Comment by koto** [comment]
We had a bit of discussions internally. It generally looks great! 

That said, an issue came up with `TrustedHTML.fromLiteral` type. While not problematic on its own, it does enable an error-prone practice of operating on HTML ignoring that HTML has multiple contexts. For example ``TrustedHTML.fromLiteral`<div foo="` `` is normalized when added to the DOM, but before that it's error-prone to use it.

In particular, concatenating TrustedHTMLs is not safe anymore, while it's reasonable to assume someone would (and thus, would shoot themselves in the foot). For example, we generally [allow concatenating](https://github.com/google/safevalues/blob/main/src/builders/html_builders.ts#L62-L64) of our safe HTMLs.

There is a simple and relatively unobtrusive change we could add to address that. In `TrustedHTML.fromLiteral`, invoke a DOM parser on the input (with, say, `<div>` node as a context), and create an instance wrapping a string serialization of the result, obtaining a well-balanced TrustedHTML as a result:

```javascript
TrustedHTML.fromLiteral`<div foo>`.toString() // <div foo=""></div>
```

It's easy to polyfill with a DOMParser. It also opens up a potential internal performance optimization in browsers, that could cache the document fragment it used for parsing, and reused that when working with a sink. 

For b/w compatibility, this would not apply to existing policy-based TrustedHTMLs. There the user has full control over the types. We might need another function for creating a full document HTML if that's needed (if someone does ``TrustedHTML.fromLiteral`<html><body>....` `` we would normally just discard the body), but apart from that I see little downsides, given that this is a new function anyway.  @shhnjk, wdyt?

**Comment by domenic** [comment]
If you go that direction, use template instead of div as the wrapper; it's canonical for these sorts of context-free scenarios.

**Comment by shhnjk** [comment]
Got it. That sounds good to me :)

**Comment by shhnjk** [comment]
Any plan to merge [the PR](https://github.com/w3c/webappsec-trusted-types/pull/350) and start an implementation?

**Comment by koto** [comment]
We still have some thoughts on the TrustedHTML variant, but I think the other two types are uncontroversial, the delay is mostly because I want to write the polyfill too.

**Comment by ai-completed** [comment]
<!DOCTYPE html>
<html lang="en">
<head>
    <meta charset="UTF-8">
    <meta http-equiv="Content-Security-Policy" content="default-src 'self';">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <title>Trusted Types Implementation</title>
    <style>
        .error {
            color: red;
        }
    </style>
</head>
<body>

<h2>Trusted Types Test</h2>
<textarea id="userInput" rows="4" cols="50" placeholder="Enter your script..."></textarea>
<button onclick="executeScript()">Execute</button>
<p id="output" class="error"></p>

<script>
    // Define a comprehensive TrustedScript policy
    const disallowedSubstrings = ['<script>', 'eval(', 'window.', 'document.', 'fetch', 'XMLHttpRequest', 'import'];
    let opaqueScriptPolicy = trustedTypes.createPolicy('comprehensivePolicy', {
        createScript: opaqueScript => {
            for (let substring of disallowedSubstrings) {
                if (opaqueScript.includes(substring)) {
                    throw new Error(`Disallowed substring detected: ${substring}`);
                }
            }
            return opaqueScript;
        },
    });

    function sanitizeInput(input) {
        const sanitized = input.replace(/</g, '&lt;').replace(/>/g, '&gt;');
        return sanitized;
    }

    function validateAndSanitize(input) {
        if (typeof input !== 'string' || input.length > 1000) {
            throw new Error("Invalid input.");
        }
        return sanitizeInput(input);
    }

    function areAllTrustedScripts(args) {
        return args.every(arg => trustedTypes.isScript(arg));
    }

    function TrustedFunctionConstructor(...args) {
        if (!areAllTrustedScripts(args)) {
            throw new Error("All arguments must be TrustedScript instances");
        }
        return new Function(...args);
    }

    function executeScript() {
        const userInputElem = document.getElementById('userInput');
        const outputElem = document.getElementById('output');

        try {
            const userInput = userInputElem.value;
            const validatedInput = validateAndSanitize(userInput);
            const opaqueExpr = opaqueScriptPolicy.createScript(validatedInput);
            TrustedFunctionConstructor(opaqueExpr)();
            outputElem.textContent = "Script executed successfully!";
            outputElem.classList.remove('error');
        } catch (error) {
            outputElem.textContent = `Error: ${error.message}`;
            outputElem.classList.add('error');
        }
    }
</script>

</body>
</html>
